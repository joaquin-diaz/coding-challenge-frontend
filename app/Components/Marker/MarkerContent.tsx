import * as React from 'react';
import styled from '@emotion/styled';

interface ContainerProps {
  hidden: boolean;
}

const Container = styled.div<ContainerProps>`
  position: absolute;
  background: white;
  box-shadow: 0 1px 6px 0 rgba(32, 33, 36, 0.28);
  box-sizing: border-box;

  bottom: 35px;
  width: 200px;
  left: -100px;
  right: 0;
  padding: 10px 20px;
  border-radius: 5px;
  font-size: 15px;

  ${props => (props.hidden ? 'opacity: 0;' : 'opacity: 1;')}
`;

interface MarkerContentProp {
  title: string;
  address: string;
  hidden: boolean;
}

const MarkerContent: React.FC<MarkerContentProp> = props => {
  return (
    <Container hidden={props.hidden}>
      {props.title} <br /> {props.address}
    </Container>
  );
};

export default MarkerContent;
