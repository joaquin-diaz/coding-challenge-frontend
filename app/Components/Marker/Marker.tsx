import * as React from 'react';
import styled from '@emotion/styled';
import styles from '../../constants/styles';

import MarkerContent from './MarkerContent';
import Film from '../../types/Film';

const { useState } = React;

const Container = styled.div`
  position: relative;
`;

const StyledMarker = styled.div`
  position: absolute;
  left: -20px;
  bottom: 0px;

  border-radius: 50% 50% 50% 0;
  border: 4px solid ${styles.colors.primary};
  width: 20px;
  height: 20px;
  transform: rotate(-45deg);

  &:after {
    position: absolute;
    content: '';
    width: 10px;
    height: 10px;
    border-radius: 50%;
    top: 50%;
    left: 50%;
    margin-left: -5px;
    margin-top: -5px;
    background-color: ${styles.colors.primary};
  }

  &:hover {
    cursor: pointer;
  }
`;

interface MarkerProps {
  lat: number;
  lng: number;
  film: Film;
  onClick: (film: Film) => void;
}

const Marker: React.FC<MarkerProps> = props => {
  const { film, onClick } = props;
  const [hovered, onHover] = useState<boolean>(false);
  const onMouseEnter = () => {
    onHover(true);
  };
  const onMouseLeave = () => {
    onHover(false);
  };
  const handleOnClick = () => {
    onClick(film);
  };

  return (
    <Container onClick={handleOnClick}>
      <StyledMarker onMouseEnter={onMouseEnter} onMouseLeave={onMouseLeave} />
      <MarkerContent title={film.title} address={film.address} hidden={!hovered} />
    </Container>
  );
};

export default Marker;
