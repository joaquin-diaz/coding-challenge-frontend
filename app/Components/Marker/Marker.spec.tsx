import * as React from 'react';

import Marker from './Marker';
import { mount } from 'enzyme';
import { fakeFilm } from '../../utils/test.fixtures';
import MarkerContent from './MarkerContent';

const noop = () => {};

describe('<Marker />', () => {
  it('should render', () => {
    const wrapper = mount(<Marker film={fakeFilm} lat={123} lng={456} onClick={noop} />);
    const markerContent = wrapper.find(MarkerContent);

    expect(wrapper).toMatchSnapshot();
    expect(markerContent.props()).toEqual({
      address: 'Epic Roasthouse (399 Embarcadero)',
      hidden: true,
      title: '180',
    });
  });

  it('should trigger hover', () => {
    const wrapper = mount(<Marker film={fakeFilm} lat={123} lng={456} onClick={noop} />);
    const marker = wrapper.find('div').at(1);

    expect(marker.prop('onMouseEnter')).toBeDefined();
    expect(marker.prop('onMouseLeave')).toBeDefined();
    marker.simulate('mouseEnter');

    const markerContent = wrapper.find(MarkerContent);
    expect(markerContent.length).toEqual(1);
    expect(markerContent.prop('hidden')).toEqual(false);
  });

  it('should trigger click', () => {
    const onClickMock = jest.fn();

    const wrapper = mount(<Marker film={fakeFilm} lat={123} lng={456} onClick={onClickMock} />);
    wrapper.simulate('click');

    expect(onClickMock).toBeCalled();
  });
});
