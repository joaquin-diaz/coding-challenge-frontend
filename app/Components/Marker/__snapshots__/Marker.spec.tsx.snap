// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Marker /> should render 1`] = `
.emotion-2 {
  position: relative;
  cursor: pointer;
}

.emotion-0 {
  position: absolute;
  left: -20px;
  bottom: 0px;
  border-radius: 50% 50% 50% 0;
  border: 4px solid #3f51b5;
  width: 20px;
  height: 20px;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
}

.emotion-0:after {
  position: absolute;
  content: '';
  width: 10px;
  height: 10px;
  border-radius: 50%;
  top: 50%;
  left: 50%;
  margin-left: -5px;
  margin-top: -5px;
  background-color: #3f51b5;
}

.emotion-1 {
  background: white;
  box-shadow: 0 1px 6px 0 rgba(32,33,36,0.28);
  box-sizing: border-box;
  border-radius: 5px;
  position: absolute;
  bottom: 35px;
  width: 200px;
  left: -100px;
  right: 0;
  padding: 10px 20px;
  font-size: 15px;
  z-index: 10;
  opacity: 0;
}

<Marker
  film={
    Object {
      "actor1": "Siddarth",
      "actor2": "Nithya Menon",
      "actor3": "Priya Anand",
      "address": "Epic Roasthouse (399 Embarcadero)",
      "coordinates": Object {
        "lat": 123,
        "lng": 456,
      },
      "director": "Jayendra",
      "productionCompany": "SPI Cinemas",
      "releaseYear": "2011",
      "title": "180",
      "writer": "Umarji Anuradha, Jayendra, Aarthi Sriram, & Suba ",
    }
  }
  lat={123}
  lng={456}
  onClick={[Function]}
>
  <Styled(div)
    onClick={[Function]}
  >
    <div
      className="emotion-2"
      onClick={[Function]}
    >
      <Styled(div)
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="emotion-0"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        />
      </Styled(div)>
      <MarkerContent
        address="Epic Roasthouse (399 Embarcadero)"
        hidden={true}
        title="180"
      >
        <Styled(div)
          hidden={true}
        >
          <div
            className="emotion-1"
            hidden={true}
          >
            180
             
            <br />
             
            Epic Roasthouse (399 Embarcadero)
          </div>
        </Styled(div)>
      </MarkerContent>
    </div>
  </Styled(div)>
</Marker>
`;
